File: src/lib/rate-limit.ts
import { createAdminClient } from "@/lib/supabase/admin";

const DAILY_QUERY_LIMIT = 50;

export interface RateLimitResult {
  allowed: boolean;
  remaining: number;
  limit: number;
  resetsAt: string;
}

export async function checkRateLimit(userId: string): Promise<RateLimitResult> {
  const admin = createAdminClient();

  // Count user messages created today
  const startOfDay = new Date();
  startOfDay.setHours(0, 0, 0, 0);

  const endOfDay = new Date();
  endOfDay.setHours(23, 59, 59, 999);

  const { count, error } = await admin
    .from("chat_messages")
    .select("*", { count: "exact", head: true })
    .eq("role", "user")
    .gte("created_at", startOfDay.toISOString())
    .lte("created_at", endOfDay.toISOString())
    // Filter by conversation ownership
    .in(
      "conversation_id",
      (
        await admin
          .from("conversations")
          .select("id")
          .eq("user_id", userId)
      ).data?.map((c) => c.id) ?? []
    );

  if (error) {
    // Fail open — don't block users if rate limit check fails
    return { allowed: true, remaining: DAILY_QUERY_LIMIT, limit: DAILY_QUERY_LIMIT, resetsAt: endOfDay.toISOString() };
  }

  const used = count ?? 0;
  const remaining = Math.max(0, DAILY_QUERY_LIMIT - used);

  return {
    allowed: remaining > 0,
    remaining,
    limit: DAILY_QUERY_LIMIT,
    resetsAt: endOfDay.toISOString(),
  };
}

File: backend/core/rate_limit.py
from datetime import datetime, timezone
from dataclasses import dataclass
from .supabase import get_supabase_admin
from .config import settings


@dataclass
class RateLimitResult:
    allowed: bool
    remaining: int
    limit: int
    resets_at: str  # ISO8601


async def check_rate_limit(user_id: str) -> RateLimitResult:
    limit = settings.daily_query_limit

    # End of today UTC
    now = datetime.now(timezone.utc)
    end_of_day = now.replace(hour=23, minute=59, second=59, microsecond=999999)
    resets_at = end_of_day.isoformat()

    try:
        supabase = get_supabase_admin()
        start_of_day = now.replace(hour=0, minute=0, second=0, microsecond=0).isoformat()

        # Single query: count user messages today via inner join on conversations
        result = (
            supabase.table("chat_messages")
            .select("id, conversations!inner(user_id)", count="exact")
            .eq("role", "user")
            .eq("conversations.user_id", user_id)
            .gte("created_at", start_of_day)
            .execute()
        )
        count = result.count or 0

        remaining = max(0, limit - count)
        return RateLimitResult(
            allowed=remaining > 0,
            remaining=remaining,
            limit=limit,
            resets_at=resets_at,
        )
    except Exception:
        # Fail open — don't block users if rate limit check fails
        return RateLimitResult(
            allowed=True,
            remaining=limit,
            limit=limit,
            resets_at=resets_at,
        )
